### DISCUSSION TITLE
How to get current sensor value after Ruletimer1 10 has expired

### DISCUSSION THREAD

**Discussion by jackpotcityco** [arendst/Tasmota#19776] [Q&A] [answered]
Hello,

I am using a `ruletimer1 10` which will have a countdown of 10 seconds:
```
Rule1 
ON VL53L1X#Distance DO IF ((%value%<=15) AND (%var2%==0)) var2 1; ruletimer1 10 ELSEIF (%value%>15) var2 0 ENDIF ENDON 
ON Rules#Timer=1 DO var1 DistanceA,[VL53L1X#Distance] ENDON
```
The above code works where the line below do execute after those 10 seconds. What I am trying to acheive is to get the current Distance from the sensor `VL53L1X` after those 10 seconds. With other word at the very time the below line do execute:
`ON Rules#Timer=1 DO var1 DistanceA,VL53L1X#Distance ENDON`

As I have understand it is not possible to use `"#"` 2 times as it is reserved for the trigger. So the `VL53L1X#Distance` only outputs that same string in the console.

How would we return the current reading of `VL53L1X` after those 10 seconds?

Thank you!

**Comment by sfromis** [comment]
You could save the value in a var in the first of your two rules, giving you the distance as quickly as it updates, and use the var in the 2nd. This is not exactly 10 seconds then, but last value you got.

Note that if you want to do things like programming, I can strongly recommend switching to a device based on ESP32, where you have the Berry programming language integrated with Tasmota. While rules are fine for basic control stuff, it does take a lot of quirks to construct logic and flows.

**Comment by jackpotcityco** [comment]
> You could save the value in a var in the first of your two rules, giving you the distance as quickly as it updates, and use the var in the 2nd. This is not exactly 10 seconds then, but last value you got.

Yes that was what I was able to do also before as you mentioned. The only problem is that the value would be 10 seconds ago when the Timer reaches Zero. It would be essential to have the current value when the Timer reach Zero.

I see, this is what I would like to get, an `ESP32` board then later. This `ESP8266EX` board is my first "try-out-things" board. I heard about Berry programming language also which seems interesting.

At the moment with the `ESP8266EX` board, it would not be possible to acheive this functionality, in the Console with Rules only?

**Reply by sfromis** [reply, marked as answer]
No, it would not be 10 seconds ago, as the trigger will fire for every sensor cycle. Of course you should not do it inside the `IF`. Maybe simpler with a separate simple rule to pick up the temperature.

And it is not the right question to ask about **possible**, but how hard a time you want to have to learn the contortional thinking to get the fullest out of rules. There are a **lot** of quirky things in how it works, and no helpful messages when you mess up. Rules work well for simple controls, but when you start to want to use multi-factor logic, things gets harder and you have to spend time on getting *good* about how to bend over backwards in the right way.

**Reply by sfromis** [reply]
On ESP8266, you also have https://tasmota.github.io/docs/Scripting-Language/ which is more powerful than rules, and a completely different learning curve to climb to make sense of.

**Reply by jackpotcityco** [reply]
Yes, it seems like alot of quirks doing that. Berry language and `ESP32` seems like a better choice for this. I should buy an `ESP32 `and learn the Berry language. I was looking at it now and seems quite "easy" to learn, as I am used to C# and a little python.

I will not try to get the reading after those 10 seconds but will wait until I have an ESP32 instead for this. I think I ment with the code below, it do save the distance value 15.0 at the time when the `RuleTimer1 `starts in `var3` which then would be 10 seconds ago. I only let the Rule1 execute once also because I set the flag variable: `var2 1` and `var2 0`

```
Rule1 
ON VL53L1X#Distance DO IF ((%value%<=15) AND (%var2%==0)) var2 1; ruletimer1 10; var3 %value% ELSEIF (%value%>15) var2 0; var3 -1 ENDIF ENDON 
ON Rules#Timer=1 DO var1 DistanceA,%var3% ENDON

```
![image](https://github.com/arendst/Tasmota/assets/47510843/19d14652-2d91-4bbf-9184-0c23b26669cf)

Thank you for your help!

**Reply by sfromis** [reply]
No, it would **not** be 10 seconds ago, if you do it outside of your `IF`. Then you could save the value for every of your messages in the screenshot, with intervals of less than 1 second.

And Berry does have some similarities with Python, but not with more advanced features like list comprehension.

**Reply by jackpotcityco** [reply]
> No, it would not be 10 seconds ago, if you do it outside of your IF. Then you could save the value for every of your messages in the screenshot, with intervals of less than 1 second.

Yes ofcourse, sorry, now I see what you ment there. This I could do, that is true as it continue to update the sensor value while the Timer1 is counting down there.

ESP32 and Berry seems very interesting and the way to go. It reminds me a bit of python as you say which is really nice.
I will get one of those ESP32 boards soon as it also is a more powerful board aswell. 

Thank you again for your help!
